dev/connect-jvm-client-mima-check

```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env bash
+
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+set -o pipefail
+set -e
+
+# Go to the Spark project root directory
+FWDIR="$(cd "`dirname "$0"`"/..; pwd)"
+cd "$FWDIR"
+export SPARK_HOME=$FWDIR
+echo $SPARK_HOME
+
+if [[ -x "$JAVA_HOME/bin/java" ]]; then
+  JAVA_CMD="$JAVA_HOME/bin/java"
+else
+  JAVA_CMD=java
+fi
+
+rm -f .connect-mima-check-result
+
+TOOLS_CLASSPATH="$(build/sbt -DcopyDependencies=false "export tools/fullClasspath" | grep jar | tail -n1)"
+CORE_CLASSPATH="$(build/sbt -DcopyDependencies=false "export core/fullClasspath" | grep jar | tail -n1)"
+CATALYST_CLASSPATH="$(build/sbt -DcopyDependencies=false "export catalyst/fullClasspath" | grep jar | tail -n1)"
+
+echo "Build sql module ..."
+build/sbt sql/package
+echo "Build connect-client-jvm module ..."
+build/sbt connect-client-jvm/assembly
+
+echo "Do connect-client-jvm module mima check ..."
+
+$JAVA_CMD \
+  -Xmx2g \
+  -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.util.jar=ALL-UNNAMED \
+  -cp "$TOOLS_CLASSPATH:$CORE_CLASSPATH:$CATALYST_CLASSPATH" \
+  org.apache.spark.tools.CheckConnectJvmClientCompatibility
+
+ERRORS=$(grep ERROR .connect-mima-check-result | tail -n1)
+
+if test ! -z "$ERRORS"; then
+  cat .connect-mima-check-result
+  echo -e "connect-client-jvm module mima check failed."
+  echo -e "Exceptions to binary compatibility can be added in tools/CheckConnectJvmClientCompatibility.scala"
```

Comment:
Can we keep the file `CheckConnectJvmClientCompatibility.scala` inside `client/jvm` and make the tools depends on the client `test-jar`? Then we do not need to copy past the `findJar` and `recursiveListFiles` methods. And also we can keep the code closer to the client?

Reply:
According to the previous habit, the `tools` module is independent, and it does not rely on any other Spark module. I do not want to break this. 

In addition, the `compatibility API tests: Dataset` related issues I mentioned in another [pr](https://github.com/apache/spark/pull/39712#discussion_r1118448264), put it inside the client module may cause unexpected class loading.

If we are sure we don't need to check `compatibility API tests: Dataset` or there are other ways to solve it, I can try to put `CheckConnectJvmClientCompatibility` back into the client module.

Reply:
Another one https://github.com/apache/spark/pull/40213 keep `CheckConnectJvmClientCompatibility ` in `client/jvm`

Reply:
Hi, I vote for change https://github.com/apache/spark/pull/40213 Thanks!